File: systems-for-design-and-implementation/labs/lab2/LibraryClient/src/library/client/gui/BooksTerminalAllAvailable.java
package library.client.gui;

import library.model.Book;

import javax.swing.table.AbstractTableModel;
import java.util.ArrayList;
import java.util.List;

public class BooksTerminalAllAvailable extends AbstractTableModel
{
    private List<Book> books = new ArrayList<Book>();
    private String[]   cols  = {"Id", "Title", "Author", "Availability"};

    @Override
    public int getRowCount()
    {
        return books.size();
    }

    @Override
    public String getColumnName(int column)
    {
        return cols[column];
    }

    @Override
    public int getColumnCount()
    {
        return cols.length;
    }

    @Override
    public Object getValueAt(int rowIndex, int columnIndex)
    {
        switch (columnIndex)
        {
            case 0:
                return books.get(rowIndex).getId();
            case 1:
                return books.get(rowIndex).getTitle();
            case 2:
                return books.get(rowIndex).getAuthor();
            case 3:
                return books.get(rowIndex).canBorrow() ? books.get(rowIndex).getStockAvailable() : "Borrowed :(";
        }

        return null;
    }

    public void updateBook(Book new_book)
    {
        // find book
        int found_index = -1;
        boolean loop = true;
        for (int i = 0; loop && i < books.size(); i++)
        {
            Book current_book = books.get(i);
            if (new_book.getId() == current_book.getId())
            {
                found_index = i;
                loop = false;
            }
        }

        if (found_index != -1)
        {
            // see if new book we can borrow, if not, remove it, otherwise update it
            if (new_book.canBorrow())
            {
                books.set(found_index, new_book);
            }
            else
            {
                books.remove(found_index);
            }
        }
        else // add
        {
            books.add(new_book);
        }

        fireTableDataChanged();
    }

    /**
     * Setter for property 'books'.
     *
     * @param books Value to set for property 'books'.
     */
    public void setBooks(List<Book> books)
    {
        this.books = books;
        fireTableDataChanged();
    }

    public Book get(int index)
    {
        return books.get(index);
    }
}

File: systems-for-design-and-implementation/labs/lab3/LibraryClient/src/chat/client/gui/BooksTerminalAllAvailable.scala
package chat.client.gui

import java.util
import javax.swing.table.AbstractTableModel

import library.model.Book

class BooksTerminalAllAvailable extends AbstractTableModel
{
    private val cols: Array[String] = Array("Id", "Title", "Author", "Availability")
    private var books: util.List[Book] = new util.ArrayList[Book]

    override def getRowCount: Int =
    {
        books.size
    }

    override def getColumnName(column: Int): String =
    {
        cols(column)
    }

    override def getColumnCount: Int =
    {
        cols.length
    }

    // sigh https://biomunky.wordpress.com/2011/02/02/scala-swing-tables/
    override def getValueAt(rowIndex: Int, columnIndex: Int): AnyRef =
    {
        columnIndex match
        {
            case 0 =>
                return books.get(rowIndex).getId.asInstanceOf[AnyRef]
            case 1 =>
                return books.get(rowIndex).getTitle
            case 2 =>
                return books.get(rowIndex).getAuthor
            case 3 =>
                return if (books.get(rowIndex).canBorrow)
                {
                    books.get(rowIndex).getStockAvailable.asInstanceOf[AnyRef]
                }
                else
                {
                    "Borrowed :("
                }
        }
        null
    }

    def updateBook(new_book: Book)
    {
        var found_index: Int = -1
        var loop: Boolean = true
        var i: Int = 0
        while (loop && i < books.size)
        {
            val current_book: Book = books.get(i)
            if (new_book.getId == current_book.getId)
            {
                found_index = i
                loop = false
            }

            i += 1
        }
        if (found_index != -1)
        {
            if (new_book.canBorrow)
            {
                books.set(found_index, new_book)
            }
            else
            {
                books.remove(found_index)
            }
        }
        else
        {
            books.add(new_book)
        }
        fireTableDataChanged()
    }

    def setBooks(books: util.List[Book])
    {
        this.books = books
        fireTableDataChanged()
    }

    def get(index: Int): Book =
    {
        books.get(index)
    }
}
